### DISCUSSION TITLE
configure to avoid http server timeout

### DISCUSSION THREAD

**Discussion by slillo** [vercel/next.js#11920] [Help]
I've a nextjs project proxied by nginx. I'm consuming a rest api served by nextjs itself that takes more than 2 minutes. Running locally without proxy in dev mode I've no problem, but in production I get error 504 timeout by nginx so I added some additional directives:

    proxy_connect_timeout 600s;
    proxy_read_timeout 600s;
    proxy_send_timeout 600s;   
    fastcgi_send_timeout 600s;
    fastcgi_read_timeout 600s;

Then the error code becomes 502. Looking at proxy_error_log file the error is
upstream prematurely closed connection while reading response header from upstream.

In my project I'm not using express or any other server explicitly so I don't know how to set http server timeout. Any way to set timeout from next.config.js file or globally for node js?

**Comment by Uranbold** [comment]
How did you resolve this?

**Reply by slillo** [reply]
I've not resolved. I had to rewrite my code in order to avoid tasks longer than 2 minutes.

**Comment by secret104278** [comment]
Hi, I recently bump into same issue. After dig in into next.js and node.js, I found this issue only occur in node12 but not in node14. This because default timeout in [node12](https://nodejs.org/docs/latest-v12.x/api/http.html#http_server_settimeout_msecs_callback) is 2 minutes and there is no timeout in [node14](https://nodejs.org/docs/latest-v14.x/api/http.html#http_request_settimeout_timeout_callback)

The only way to get around this in next.js with node12 is using [`--http-server-default-timeout`](https://nodejs.org/docs/latest-v12.x/api/cli.html#cli_http_server_default_timeout_milliseconds) by something like `NODE_OPTIONS='--http-server-default-timeout=0' next start`

Hope this help anyone

**Comment by Psychosynthesis** [comment]
Hmm ... Using the `NODE_OPTIONS='--http-server-default-timeout=0'` looks like a real crutch... 

_By the way, if someone sets a non-zero timeout for their build in this way, do not forget that it is set in milliseconds._

For example, if there are some node-scripts in the assembly that are executed by the "run build" or "run dev" commands - the "scripts" section in package.json can swell to incredible length, which looks as miserable as possible.

**Reply by Enigma10** [reply]
@Psychosynthesis  @secret104278  I am still facing this error on nodejs14 also. But we are not having a custom express server and we are making API call in getserverprops. Our app is deployed on AWS EC2 instance.

**Reply by Enigma10** [reply]
@Psychosynthesis Do I still have to make these ngix changes? if my API is taking a long time. I am sorry to ask this question. But I am not only a frontend developer and trying to find issue.

`proxy_connect_timeout 600s;
proxy_read_timeout 600s;
proxy_send_timeout 600s;   
fastcgi_send_timeout 600s;
fastcgi_read_timeout 600s;`

**Reply by Psychosynthesis** [reply]
@Enigma10 i don't know about Nginx... Option which i was talking about
`NODE_OPTIONS='--http-server-default-timeout=0'`
Is not from Nginx config. This option pass to nodejs when your app started from terminal

I honestly am not good at configuring nginx, but in my little experience with it, there have not yet been such cases when i had to additionally configure its timeouts, you may not have to.

**Reply by Psychosynthesis** [reply]
@Enigma10 Even if you do not use a custom server, NextJS somehow has its own server side, so I think that you still need to try the option `--http-server-default-timeout` first, and only then torture Nginx.
